[Spark-7511][MLLIB] pyspark ml seed param should be random by default or 42 is quite funny but not very random

holdenk opened the pull request.

holdenk committed:
Make the seed random for HasSeed in python

holdenk committed:
Simplify the generated code to only include set default if there is a default rather than having None is note None in the generated code

holdenk committed:
Make seed default to random in more places

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32673 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32673/consoleFull) for   PR 6139 at commit [`0e3797e`](https://github.com/apache/spark/commit/0e3797e48c92791f529f1879336a795acbac3502).

SparkQA commented:
[Test build #32673 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32673/consoleFull) for   PR 6139 at commit [`0e3797e`](https://github.com/apache/spark/commit/0e3797e48c92791f529f1879336a795acbac3502).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32673/
Test FAILed.

JoshRosen commented:
![random-number-generator-dilbert-comic](https://cloud.githubusercontent.com/assets/50748/7625867/7994401a-f9b2-11e4-99b8-75436779cd05.jpg)

holdenk committed:
pep8 fixes

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32688 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32688/consoleFull) for   PR 6139 at commit [`65eba21`](https://github.com/apache/spark/commit/65eba21789884d494dda532eefe2345b028b91e3).

holdenk commented:
@JoshRosen 
![image](https://cloud.githubusercontent.com/assets/59893/7627734/3cd33cae-f9cf-11e4-87a3-a482caba62ed.png)
 :)

mengxr commented on python/pyspark/ml/feature.py:
```diff
@@ -790,7 +790,7 @@ class Word2Vec(JavaEstimator, HasStepSize, HasMaxIter, HasSeed, HasInputCol, Has
+                 seed=None, inputCol=None, outputCol=None):
         """
         __init__(self, vectorSize=100, minCount=5, numPartitions=1, stepSize=0.025, maxIter=1,
                  seed=42, inputCol=None, outputCol=None)
```
We also need to update the docstring to keep the same signature.

holdenk replied:
Good point, I originally didn't because I though they might have been doctests (and I wanted to keep the tests predictable), but that doesn't seem to be the case. I'll update these :)

mengxr commented on python/pyspark/ml/param/_shared_params_code_gen.py:
```diff
@@ -115,7 +118,7 @@ def get$Name(self):
         ("numFeatures", "number of features", None),
         ("checkpointInterval", "checkpoint interval (>= 1)", None),
-        ("seed", "random seed", None),
+        ("seed", "random seed", "random.randint(0, sys.maxsize)"),
```
I agree that setting the default seed to 42 globally is quite funny. But I think it is nice to have a fixed seed for each component, so it is easy to reproduce the result. For example, could you send the seed based on the hash value of the class name?

holdenk replied:
So the original bug mentioned that in Scala it was done randomly. I think for testing using a fixed seed in the constructor is a good way to ensure reproduceability. I'm fine with replacing this with a hash of the class or something if we think that would be better, but I think it probably makes sense to do the same thing between Python & Scala.

jkbradley replied:
Yes, in Scala, we use a random seed each time.  (See the code gen file for HasSeed, or several examples in spark.mllib.)  I think that's the best behavior: By default, we use a random seed.  If users want reproducible behavior, they should set the seed manually.

mengxr commented on python/pyspark/ml/tests.py:
```diff
@@ -112,14 +112,15 @@ def test_pipeline(self):
+    def __init__(self, seed=None):
         super(TestParams, self).__init__()
         self._setDefault(maxIter=10)
+        self._set(seed=seed)
```
If the input `seed` is `None`, this will set `seed` to `None`.

holdenk replied:
Yes, that is the desired effect. If in the kwargs it is set to one then we use the default. This also matches the behaviour taken in the classes that use hasSeed (although they do wrap it through a few extra levels of function calls I left off here for test simplicity).

jkbradley replied:
It should probably do what other subclasses of Params do.  E.g., look at Word2Vec in features.py:
```
        kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)
```
This way, Param "seed" will only be explicitly set in self._paramMap if it is specified by the user.  We want to be able to distinguish between users specifying a parameter vs. not.

mengxr commented on python/pyspark/ml/tests.py:
```diff
@@ -153,10 +155,26 @@ def test_params(self):
+        noSeedSpecd = TestParams()
+        withSeedSpecd = TestParams(seed=42)
+        # Check that we no longer use 42 as the magic number
+        self.assertNotEqual(noSeedSpecd.getSeed(), 42)
```
This is a flaky test, though with a tiny tiny probability.

holdenk replied:
It is. I can make sure that 42 is never accepted as a random number or try this test k times, but I think 1/2^32 flakiness is maybe ok?

mengxr commented:
@holdenk @JoshRosen Thanks for sharing!

SparkQA commented:
[Test build #32688 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32688/consoleFull) for   PR 6139 at commit [`65eba21`](https://github.com/apache/spark/commit/65eba21789884d494dda532eefe2345b028b91e3).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32688/
Test PASSed.

mengxr commented:
@holdenk Please add `[MLLIB]` to the PR title so it gets listed under "MLlib" on https://spark-prs.appspot.com/#mllib.

holdenk commented:
@mengxr Thanks for revewing this! I'll update the PR title :)

jkbradley commented:
Haha, I like this PR so far : )

holdenk committed:
update docstring seeds to none and some other default seeds from 42

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32740/
Test FAILed.

holdenk commented:
Seems like its failing from a git issue. @AmplabJenkins retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32743 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32743/consoleFull) for   PR 6139 at commit [`bcdf3c9`](https://github.com/apache/spark/commit/bcdf3c9792473876e4e481c00045d45ce74d00d6).

SparkQA commented:
[Test build #32743 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32743/consoleFull) for   PR 6139 at commit [`bcdf3c9`](https://github.com/apache/spark/commit/bcdf3c9792473876e4e481c00045d45ce74d00d6).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32743/
Test FAILed.

SparkQA commented:
[Test build #806 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/806/consoleFull) for   PR 6139 at commit [`bcdf3c9`](https://github.com/apache/spark/commit/bcdf3c9792473876e4e481c00045d45ce74d00d6).

SparkQA commented:
[Test build #806 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/806/consoleFull) for   PR 6139 at commit [`bcdf3c9`](https://github.com/apache/spark/commit/bcdf3c9792473876e4e481c00045d45ce74d00d6).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

holdenk committed:
merge in master

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32758 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32758/consoleFull) for   PR 6139 at commit [`ce90ec8`](https://github.com/apache/spark/commit/ce90ec87e5a16ec99805d26204c3334f946275a1).

SparkQA commented:
[Test build #32758 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32758/consoleFull) for   PR 6139 at commit [`ce90ec8`](https://github.com/apache/spark/commit/ce90ec87e5a16ec99805d26204c3334f946275a1).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32758/
Test FAILed.

holdenk committed:
Style fixes

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32761 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32761/consoleFull) for   PR 6139 at commit [`e1b947f`](https://github.com/apache/spark/commit/e1b947fdcc23074cbae7f57298bff111c58d7fd0).

SparkQA commented:
[Test build #32761 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32761/consoleFull) for   PR 6139 at commit [`e1b947f`](https://github.com/apache/spark/commit/e1b947fdcc23074cbae7f57298bff111c58d7fd0).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32761/
Test FAILed.

jkbradley commented:
That test failure looks kind of valid.  I'll look into it...

holdenk committed:
specify the seed to randomforestregressor test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

holdenk commented:
@jkbradley: I think its just that the seed wasn't specified in that test, I've gone ahead and specified the seed for that test and we should be good to go (works locally well see about what Jenkins has to say).

SparkQA commented:
[Test build #32792 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32792/consoleFull) for   PR 6139 at commit [`31cd96f`](https://github.com/apache/spark/commit/31cd96f7f2b909b846f849a4400244bb933d94dd).

jkbradley commented:
I had been thinking that the example should work no matter what seed is used, but perhaps it's randomness from subsampling the data.  In that case, the implementation is correct...just not robust for tiny datasets (which may be OK for Spark users).  You're probably right that specifying the seed in the test is fine for now.

holdenk commented:
Cool, let me know if I should do something else with the RandomForestRegressor :)

jkbradley commented:
Hopefully good for now--thanks!

SparkQA commented:
[Test build #32792 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32792/consoleFull) for   PR 6139 at commit [`31cd96f`](https://github.com/apache/spark/commit/31cd96f7f2b909b846f849a4400244bb933d94dd).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32792/
Test FAILed.

holdenk commented:
I suspect the streaming test failure is unrelated. @AmplabJenkins retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #32840 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32840/consoleFull) for   PR 6139 at commit [`31cd96f`](https://github.com/apache/spark/commit/31cd96f7f2b909b846f849a4400244bb933d94dd).

SparkQA commented:
[Test build #32840 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32840/consoleFull) for   PR 6139 at commit [`31cd96f`](https://github.com/apache/spark/commit/31cd96f7f2b909b846f849a4400244bb933d94dd).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/32840/
Test PASSed.

jkbradley commented:
@holdenk After discussing more with @mengxr I think we'll go with using a fixed random seed, but one which is generated based on the name of the class.  The motivation is to make it easier to reproduce things, but to have it differ per class to keep things a little more random.  Could you please change the code gen file so it sets the seed based on some hash of the class name?  I'll make a separate JIRA for us to do this in Scala.  Thanks!

holdenk commented:
Sure I'll give that a shot tomorrow :)

holdenk committed:
Merge branch 'master' into SPARK-7511-pyspark-ml-seed-param-should-be-random-by-default-or-42-is-quite-funny-but-not-very-random

holdenk committed:
switch default seed to hash of type of self

holdenk committed:
make sure different classes have different results

holdenk committed:
fix test and regenerate base

holdenk committed:
its the __name__ method, also treat None values as not set

jkbradley commented:
Thanks!

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

holdenk committed:
merge in master

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33058 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33058/consoleFull) for   PR 6139 at commit [`8616751`](https://github.com/apache/spark/commit/86167513d316dfae35f1bd048e168761bfaa5d9f).

SparkQA commented:
[Test build #33058 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33058/consoleFull) for   PR 6139 at commit [`8616751`](https://github.com/apache/spark/commit/86167513d316dfae35f1bd048e168761bfaa5d9f).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33058/
Test FAILed.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33057/
Test FAILed.

holdenk committed:
Add some missing newlines for pep8 style

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33065 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33065/consoleFull) for   PR 6139 at commit [`926165a`](https://github.com/apache/spark/commit/926165a84b6b4fa9ab24e18ff84cc4592cc9c641).

SparkQA commented:
[Test build #33065 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33065/consoleFull) for   PR 6139 at commit [`926165a`](https://github.com/apache/spark/commit/926165a84b6b4fa9ab24e18ff84cc4592cc9c641).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33065/
Test FAILed.

jkbradley commented on python/pyspark/ml/param/__init__.py:
```diff
@@ -147,10 +147,10 @@ def hasParam(self, paramName):
         """
         param = self._resolveParam(param)
-        if param in self._paramMap:
+        if param in self._paramMap and self.paramMap[param]:
```
Was this merged incorrectly?  This line should not be modified, I believe.

holdenk replied:
This was an intentional change. If something is present in the param map but had None then we should use the default value.

jkbradley replied:
I don't think the value could be None unless the user explicitly sets it to be None, in which case we should probably assume that "None" is a valid value.

jkbradley commented on python/pyspark/ml/param/shared.py:
```diff
@@ -482,7 +459,7 @@ def __init__(self):
         #: param for If false, the algorithm will pass trees to executors to match instances with nodes. If true, the algorithm will cache node IDs for each instance. Caching can speed up training of deeper trees.
         self.cacheNodeIds = Param(self, "cacheNodeIds", "If false, the algorithm will pass trees to executors to match instances with nodes. If true, the algorithm will cache node IDs for each instance. Caching can speed up training of deeper trees.")
-
+        
```
Not needed

jkbradley commented:
@holdenk  Thanks for the update!  I think the hash is fine, but I had a few comments, including a response early in the PR to a comment from @mengxr (where I think he was right).

holdenk committed:
remove extra space

holdenk committed:
Switch the test code to behave slightly more like production code. also don't check the param map value only check for key existence

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33087 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33087/consoleFull) for   PR 6139 at commit [`140d25d`](https://github.com/apache/spark/commit/140d25d7a5023ef070f9a17a2ea6b32a1b34d15f).

holdenk committed:
Add the setParams function that is used in the real code

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33088 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33088/consoleFull) for   PR 6139 at commit [`cbad96d`](https://github.com/apache/spark/commit/cbad96d262a126a8c049573d8ea28001827014ad).

holdenk commented:
@jkbradley thanks :) I've updated based on the CR feedback and I removed the extra check for the param being not None now that the test code uses the same setParam path as the real code.

jkbradley commented:
@holdenk Thanks, glad this PR pointed out that the TestParams class didn't exactly follow the rules!  The setParams() method isn't used anywhere, so it could be removed.  (It's not overriding anything, so it won't be called from elsewhere.)  Other than that, it looks good.

SparkQA commented:
[Test build #33087 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33087/consoleFull) for   PR 6139 at commit [`140d25d`](https://github.com/apache/spark/commit/140d25d7a5023ef070f9a17a2ea6b32a1b34d15f).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33087/
Test FAILed.

SparkQA commented:
[Test build #33088 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33088/consoleFull) for   PR 6139 at commit [`cbad96d`](https://github.com/apache/spark/commit/cbad96d262a126a8c049573d8ea28001827014ad).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33088/
Test PASSed.

jkbradley commented:
@holdenk I was about to give the looks-good, but I just noticed that there are still classes implementing HasSeed which use old defaults (in classification.py and recommendation.py).  Could you please update those too?  Note: The old defaults they use match the old ones in Scala, but Scala has been changed now too.  After that, this should be ready.  Thanks a lot!

holdenk commented:
Sure thing

holdenk committed:
Fix a bunch of seeds with default values to have None as the default which will then result in using the hash of the class name

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33112 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33112/consoleFull) for   PR 6139 at commit [`2470004`](https://github.com/apache/spark/commit/2470004a47c8fa50baf9399d0c83502f2b825b2e).

jkbradley commented:
LGTM once tests pass

SparkQA commented:
[Test build #33112 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33112/consoleFull) for   PR 6139 at commit [`2470004`](https://github.com/apache/spark/commit/2470004a47c8fa50baf9399d0c83502f2b825b2e).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33112/
Test FAILed.

jkbradley commented:
Probably need to set classification.py:295 (RandomForestClassifier doc test) seed to old fixed value.  (42?)

holdenk committed:
specify old seed for doc tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33129 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33129/consoleFull) for   PR 6139 at commit [`591f8e5`](https://github.com/apache/spark/commit/591f8e584f3ad116da3fd436535cc875b8c01a2d).

SparkQA commented:
**[Test build #33129 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33129/consoleFull)**     for PR 6139 at commit [`591f8e5`](https://github.com/apache/spark/commit/591f8e584f3ad116da3fd436535cc875b8c01a2d)     after a configured wait of `150m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33129/
Test FAILed.

mengxr commented:
@tdas @davies There were timeouts in PySpark Streaming unit tests: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33129/consoleFull.

mengxr commented:
test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33146 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33146/consoleFull) for   PR 6139 at commit [`591f8e5`](https://github.com/apache/spark/commit/591f8e584f3ad116da3fd436535cc875b8c01a2d).

SparkQA commented:
[Test build #33146 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33146/consoleFull) for   PR 6139 at commit [`591f8e5`](https://github.com/apache/spark/commit/591f8e584f3ad116da3fd436535cc875b8c01a2d).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33146/
Test FAILed.

holdenk commented:
since we didn't change any Scala code failing the Scala style tests is unexpected. @AmplabJenkins retest this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #33150 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33150/consoleFull) for   PR 6139 at commit [`591f8e5`](https://github.com/apache/spark/commit/591f8e584f3ad116da3fd436535cc875b8c01a2d).

SparkQA commented:
[Test build #33150 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33150/consoleFull) for   PR 6139 at commit [`591f8e5`](https://github.com/apache/spark/commit/591f8e584f3ad116da3fd436535cc875b8c01a2d).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/33150/
Test FAILed.

SparkQA commented:
[Test build #837 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/837/consoleFull) for   PR 6139 at commit [`591f8e5`](https://github.com/apache/spark/commit/591f8e584f3ad116da3fd436535cc875b8c01a2d).

SparkQA commented:
[Test build #837 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/837/consoleFull) for   PR 6139 at commit [`591f8e5`](https://github.com/apache/spark/commit/591f8e584f3ad116da3fd436535cc875b8c01a2d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

jkbradley commented:
Merging into master and branch-1.4  (branch-1.4 is merited since this changes algorithm behavior)
@holdenk Thanks!
